046-analyse-broker-metrics-before-rolling-kafka.md

```diff
@@ -0,0 +1,93 @@
+
+## Current situation
+
+The current logic in KafkaRoller:
+- Take a list of Kafka pods and reorder them based on the readiness.
+- Roll unready pods first without considering why it could be unready.
+- Before restarting a Kafka pod, consider if it is a controller or if it has an impact on the availability.
+- If a Kafka pod does not become ready within the operational timeout (300000ms by default) after restarting, then it may force restart it without considering why it's taking a long time.
+
+The current flow of KafkaRoller
+![The current flow of KafkaRoller](images/046-kafka-roller-current-flow.png)
+
+A Kafka broker can take a long time to become ready while performing log recovery. In this case, KafkaRoller could end up force restarting the broker and this can continue indefinitely, because of the log recovery on the broker startup again. KafkaRoller currently does not have a way to know what state a broker is in before making a decision to force restart it.
+
+KafkaAgent currently collects the `kafka.server:type=KafkaServer,name=BrokerState` metric and creates a `kafka-ready` file on disk if the broker state value is greater than `3`(`RUNNING`) and not `127` (`UNKNOWN`). The Kafka readiness check passes if it finds the file on disk.
+
+## Motivation
+
+To address the issues with restarts and log recovery, KafkaRoller needs to be aware of the current broker state.  If a broker is performing log recovery, KafkaRoller should fail the reconciliation instead of force restarting it.
+
+## Proposal
+ 
+The objective of this proposal is to introduce a REST API endpoint within KafkaAgent, which exposes metrics from a Kafka broker. Furthermore, KafkaRoller would be modified so that it can query this endpoint to determine the current state of a broker, thus preventing unnecessary restarts. Additionally, KafkaRoller would allow a human operator to observe whether the log recovery process is making progress or is stuck.
+
+The following implementation details will describe behaviour of the new API in KafkaAgent and how KafkaRoller would interact with it and handle a broker that is in log recovery.
+
+### Implementation details
+
+**KafkaAgent**:
+
+- Add a web server to accept SSL connections on port 8443.
+- Use Jetty for the web server since Kafka already uses it therefore avoids an additional dependency.
+- Reuse the broker's own certificate that would be in the container for TLS authentication. 
+- Configure the server to require TLS client authentication. 
+- Expose an endpoint `v1/recovery-state` that includes the API version in the URI path.
+- Extend KafkaAgent to collect `kafka.log:type=LogManager,name=remainingLogsToRecover` and `kafka.log:type=LogManager,name=remainingSegmentsToRecover` metrics.
+- Use Jetty's native ServerHandler API to handle a `GET` request and return a JSON response that looks like this:
+```
+{
+  "brokerState": 2,
+  "remainingLogsToRecover": 123,
+  "remainingSegmentsToRecover": 456
+}
+```
```

Comment:
What would it return when it is in state `3` and is running? What is the `/recovery-state` in such case? Is it empty? Or should it be `/broker-state` and always return the state and only when the state is `2` also return the progress?

Reply:
The intention is to always return the broker state and KafkaRoller would check the value. The progress values would be empty I guess if the state is 3, for example:
```
{
  "brokerState": 3,
  "remainingLogsToRecover":0,
  "remainingSegmentsToRecover": 0
}
```
KafkaRoller would not check the progress values if the state is not equal to 2.

Reply:
Might be worth specifying it in the document. I also wonder:
* if `broker-state` is a better name for the endpoint if it essentially covers the whole broker state and not just the recovery phase
* Also, what HTTP code would it return? If it always returns 200, it would not be possible to use it as healthcheck later and we would need a separate endpoint for it. (no strong opinion on which way is better. But would be wort clarifying here.)

Reply:
Wondering if, by using `broker-state` as endpoint name we could have an optional `recovery` field in the JSON when there is the need to show the remaining logs and segments to recover instead of 0.
It would be:

```json
{
  "brokerState": 2,
  "recovery": {
       "remainingLogsToRecover": 123,
       "remainingSegmentsToRecover": 456
   }
}
```

or 

```json
{
  "brokerState": 3,
}
```

or `recoveryState` instead of `recovery`.

Reply:
> Wondering if, by using broker-state as endpoint name we could have an optional recovery field in the JSON when there is the need to show the remaining logs and segments to recover instead of 0. 

I think we would always need those fields but only use them to log if brokerState==2, to avoid sending multiple HTTP requests.

Reply:
Why multiple requests? If brokerState is 3, there is nothing to log. If brokerState is 2, you get the fields and log them. Everything in one request.

Reply:
> Yes, I was thinking it would return 200 but I'm not sure what you mean by using it as healthcheck later?

The HTTP probe works the way that it checks for HTTP status `>= 200 && < 400`. So, depending on the status codes, this endpoint might or might not be usable for a liveness probe or readiness probe based on the HTTP endpoint.

E.g. if it always returns 200 here, we would need a separate endpoint for readiness to return HTTP 200 only when state is 3 for example - which is not wrong per-se, just got to be clarified.

Reply:
For liveness and readiness probe we would probably have to create an internal non-TLS endpoint, which could then return 200 only if broker state is >3 and !=127 ?

Reply:
> Why multiple requests? If brokerState is 3, there is nothing to log. If brokerState is 2, you get the fields and log them. Everything in one request.

Ah I misunderstood it initially. So the server can include the recovery fields in the response only if the broker state is 2. 

I will clarify both points above in the proposal :)